File: test/ModularPipelines.UnitTests/Configuration/ModuleConfigurationTests.cs
using Kevlar;
using ModularPipelines.Configuration;
using ModularPipelines.Context;
using ModularPipelines.Enums;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using Moq;

namespace ModularPipelines.UnitTests.Configuration;

public class ModuleConfigurationTests
{
    private sealed class DependencyModule : Module<string>
    {
        protected internal override Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
            => Task.FromResult(string.Empty);
    }

    #region Default Tests

    [Test]
    public async Task Default_SkipCondition_IsNull()
    {
        var config = ModuleConfiguration.Default;

        await Assert.That((object?) config.SkipCondition).IsNull();
    }

    [Test]
    public async Task Default_Timeout_IsNull()
    {
        var config = ModuleConfiguration.Default;

        await Assert.That(config.Timeout).IsNull();
    }

    [Test]
    public async Task Default_RetryConfiguration_IsNull()
    {
        var config = ModuleConfiguration.Default;

        using (Assert.Multiple())
        {
            await Assert.That(config.RetryConfiguration).IsNull();
            await Assert.That(config.ResilienceShieldFactory).IsNull();
        }
    }

    [Test]
    public async Task Default_IgnoreFailuresCondition_IsNull()
    {
        var config = ModuleConfiguration.Default;

        await Assert.That(config.IgnoreFailuresCondition).IsNull();
    }

    [Test]
    public async Task Default_AlwaysRun_IsFalse()
    {
        var config = ModuleConfiguration.Default;

        await Assert.That(config.AlwaysRun).IsFalse();
    }

    [Test]
    public async Task Create_ReturnsBuilder()
    {
        var builder = ModuleConfiguration.Create();

        await Assert.That(builder).IsNotNull();
        await Assert.That(builder).IsTypeOf<ModuleConfigurationBuilder>();
    }

    #endregion

    #region WithSkipWhen Tests

    [Test]
    public async Task WithSkipWhen_ExposesOnlyComposableOverloads()
    {
        var parameterTypes = typeof(ModuleConfigurationBuilder)
            .GetMethods()
            .Where(method => method.Name == nameof(ModuleConfigurationBuilder.WithSkipWhen))
            .Select(method => method.GetParameters().Single().ParameterType)
            .ToArray();

        await Assert.That(parameterTypes).IsEquivalentTo(
        [
            typeof(Func<IModuleContext, SkipDecision>),
            typeof(Func<IModuleContext, CancellationToken, ValueTask<SkipDecision>>),
        ]);
    }

    [Test]
    public async Task WithSkipWhenAll_ExposesSyncAndAsyncGroups()
    {
        var parameterTypes = typeof(ModuleConfigurationBuilder)
            .GetMethods()
            .Where(method => method.Name == nameof(ModuleConfigurationBuilder.WithSkipWhenAll))
            .Select(method => method.GetParameters().Single().ParameterType)
            .ToArray();

        await Assert.That(parameterTypes).IsEquivalentTo(
        [
            typeof(Func<IModuleContext, SkipDecision>[]),
            typeof(Func<IModuleContext, CancellationToken, ValueTask<SkipDecision>>[]),
        ]);
    }

    [Test]
    public async Task WithSkipWhen_RepeatedCalls_OrComposeAndShortCircuit()
    {
        var evaluatedConditions = new List<string>();
        var config = ModuleConfiguration.Create()
            .WithSkipWhen(_ =>
            {
                evaluatedConditions.Add("first");
                return SkipDecision.Skip("First reason");
            })
            .WithSkipWhen(_ =>
            {
                evaluatedConditions.Add("second");
                return SkipDecision.DoNotSkip;
            })
            .Build();

        var decision = await config.SkipCondition!(Mock.Of<IModuleContext>(), CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("First reason");
            await Assert.That(evaluatedConditions).IsEquivalentTo(["first"]);
        }
    }

    [Test]
    public async Task WithSkipWhen_RepeatedCalls_SkipWhenLaterConditionMatches()
    {
        var config = ModuleConfiguration.Create()
            .WithSkipWhen(_ => SkipDecision.DoNotSkip)
            .WithSkipWhen(_ => SkipDecision.Skip("Second reason"))
            .Build();

        var decision = await config.SkipCondition!(Mock.Of<IModuleContext>(), CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("Second reason");
        }
    }

    [Test]
    public async Task WithSkipWhen_SyncCondition_ReceivesContext()
    {
        var context = Mock.Of<IModuleContext>();
        var expectedDecision = SkipDecision.Skip("Test reason");

        var config = ModuleConfiguration.Create()
            .WithSkipWhen(receivedContext =>
                ReferenceEquals(receivedContext, context) ? expectedDecision : SkipDecision.DoNotSkip)
            .Build();

        var decision = await config.SkipCondition!(context, CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("Test reason");
        }
    }

    [Test]
    public async Task WithSkipWhen_AsyncCondition_SetsSkipCondition()
    {
        var expectedDecision = SkipDecision.Skip("Async reason");

        var config = ModuleConfiguration.Create()
            .WithSkipWhen(async (_, _) =>
            {
                await Task.Delay(1).ConfigureAwait(false);
                return expectedDecision;
            })
            .Build();

        var context = Mock.Of<IModuleContext>();
        var decision = await config.SkipCondition!(context, CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("Async reason");
        }
    }

    [Test]
    public async Task WithSkipWhen_AsyncCondition_ReceivesCancellationToken()
    {
        using var cancellationTokenSource = new CancellationTokenSource();
        CancellationToken? receivedToken = null;
        var config = ModuleConfiguration.Create()
            .WithSkipWhen((_, cancellationToken) =>
            {
                receivedToken = cancellationToken;
                return ValueTask.FromResult(SkipDecision.Skip("Testing"));
            })
            .Build();

        await config.SkipCondition!(Mock.Of<IModuleContext>(), cancellationTokenSource.Token);

        await Assert.That(receivedToken).IsEqualTo(cancellationTokenSource.Token);
    }

    [Test]
    public async Task WithSkipWhenAll_AllConditionsSkip_CombinesReasons()
    {
        var config = ModuleConfiguration.Create()
            .WithSkipWhenAll(
                _ => SkipDecision.Skip("First reason"),
                _ => SkipDecision.Skip("Second reason"))
            .Build();

        var decision = await config.SkipCondition!(Mock.Of<IModuleContext>(), CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("First reason; Second reason");
        }
    }

    [Test]
    public async Task WithSkipWhenAll_StopsWhenConditionDoesNotSkip()
    {
        var evaluatedConditions = new List<string>();
        var config = ModuleConfiguration.Create()
            .WithSkipWhenAll(
                _ =>
                {
                    evaluatedConditions.Add("first");
                    return SkipDecision.DoNotSkip;
                },
                _ =>
                {
                    evaluatedConditions.Add("second");
                    return SkipDecision.Skip("Should not be evaluated");
                })
            .Build();

        var decision = await config.SkipCondition!(Mock.Of<IModuleContext>(), CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsFalse();
            await Assert.That(evaluatedConditions).IsEquivalentTo(["first"]);
        }
    }

    [Test]
    public async Task WithSkipWhenAll_SnapshotsAsyncConditionGroups()
    {
        Func<IModuleContext, CancellationToken, ValueTask<SkipDecision>>[] conditions =
        [
            (_, _) => ValueTask.FromResult(SkipDecision.Skip("Original reason")),
        ];
        var config = ModuleConfiguration.Create()
            .WithSkipWhenAll(conditions)
            .Build();

        conditions[0] = (_, _) => ValueTask.FromResult(SkipDecision.DoNotSkip);

        var decision = await config.SkipCondition!(
            Mock.Of<IModuleContext>(),
            CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(decision.ShouldSkip).IsTrue();
            await Assert.That(decision.Reason).IsEqualTo("Original reason");
        }
    }

    [Test]
    public void WithSkipWhenAll_RejectsEmptyGroups()
    {
        var builder = ModuleConfiguration.Create();

        Assert.Throws<ArgumentException>(() =>
            builder.WithSkipWhenAll(Array.Empty<Func<IModuleContext, SkipDecision>>()));
    }

    [Test]
    public async Task Build_SnapshotsSkipConditions()
    {
        var builder = ModuleConfiguration.Create()
            .WithSkipWhen(_ => SkipDecision.DoNotSkip);
        var firstConfig = builder.Build();

        builder.WithSkipWhen(_ => SkipDecision.Skip("Second reason"));
        var secondConfig = builder.Build();

        var context = Mock.Of<IModuleContext>();
        var firstDecision = await firstConfig.SkipCondition!(context, CancellationToken.None);
        var secondDecision = await secondConfig.SkipCondition!(context, CancellationToken.None);

        using (Assert.Multiple())
        {
            await Assert.That(firstDecision.ShouldSkip).IsFalse();
            await Assert.That(secondDecision.ShouldSkip).IsTrue();
        }
    }

    #endregion

    #region WithTimeout Tests

    [Test]
    public async Task WithTimeout_SetsTimeout()
    {
        var timeout = TimeSpan.FromMinutes(5);

        var config = ModuleConfiguration.Create()
            .WithTimeout(timeout)
            .Build();

        await Assert.That(config.Timeout).IsEqualTo(timeout);
    }

    #endregion

    #region WithRetry Tests

    [Test]
    public async Task WithRetry_UsesDefaultBaseDelay()
    {
        var config = ModuleConfiguration.Create()
            .WithRetry(3)
            .Build();

        using (Assert.Multiple())
        {
            await Assert.That(config.RetryConfiguration).IsNotNull();
            await Assert.That(config.RetryConfiguration!.Count).IsEqualTo(3);
            await Assert.That(config.RetryConfiguration.BaseDelay).IsEqualTo(TimeSpan.FromMilliseconds(100));
            await Assert.That(config.RetryConfiguration.ShouldRetry).IsNull();
            await Assert.That(config.ResilienceShieldFactory).IsNull();
        }
    }

    [Test]
    public async Task WithRetry_StoresBaseDelayAndExceptionFilter()
    {
        var baseDelay = TimeSpan.FromSeconds(2);
        Func<Exception, bool> shouldRetry = exception => exception is TimeoutException;

        var config = ModuleConfiguration.Create()
            .WithRetry(5, baseDelay, shouldRetry)
            .Build();

        using (Assert.Multiple())
        {
            await Assert.That(config.RetryConfiguration).IsNotNull();
            await Assert.That(config.RetryConfiguration!.Count).IsEqualTo(5);
            await Assert.That(config.RetryConfiguration.BaseDelay).IsEqualTo(baseDelay);
            await Assert.That(config.RetryConfiguration.ShouldRetry).IsSameReferenceAs(shouldRetry);
        }
    }

    [Test]
    public async Task WithRetry_RejectsNegativeValues()
    {
        var countException = Assert.Throws<ArgumentOutOfRangeException>(
            () => ModuleConfiguration.Create().WithRetry(-1));
        var delayException = Assert.Throws<ArgumentOutOfRangeException>(
            () => ModuleConfiguration.Create().WithRetry(1, TimeSpan.FromMilliseconds(-1)));

        using (Assert.Multiple())
        {
            await Assert.That(countException.ParamName).IsEqualTo("count");
            await Assert.That(delayException.ParamName).IsEqualTo("baseDelay");
        }
    }

    [Test]
    public async Task Advanced_WithShield_Direct_SetsResilienceShieldFactory()
    {
        var shield = Shield.Retry(0);

        var config = ModuleConfiguration.Create()
            .Advanced
            .WithShield(shield)
            .Build();

        await Assert.That(config.ResilienceShieldFactory).IsNotNull();

        var context = Mock.Of<IModuleContext>();
        var result = config.ResilienceShieldFactory!(context);

        await Assert.That(result).IsEqualTo(shield);
    }

    [Test]
    public async Task Advanced_WithShield_Factory_SetsResilienceShieldFactory()
    {
        var shield = Shield.Retry(0);

        var config = ModuleConfiguration.Create()
            .Advanced
            .WithShield(_ => shield)
            .Build();

        await Assert.That(config.ResilienceShieldFactory).IsNotNull();

        var context = Mock.Of<IModuleContext>();
        var result = config.ResilienceShieldFactory!(context);

        await Assert.That(result).IsEqualTo(shield);
    }

    [Test]
    public async Task StandardConfigurationSurface_DoesNotExposeKevlarTypes()
    {
        var publicSurfaceTypes = typeof(ModuleConfigurationBuilder)
            .GetMethods()
            .SelectMany(method => method.GetParameters()
                .Select(parameter => parameter.ParameterType)
                .Append(method.ReturnType))
            .Concat(typeof(ModuleConfiguration)
                .GetProperties()
                .Select(property => property.PropertyType));

        await Assert.That(publicSurfaceTypes.Any(ContainsKevlarType)).IsFalse();
    }

    [Test]
    [Arguments(0, 100)]
    [Arguments(1, 200)]
    public async Task RetryDelayCalculator_AddsBoundedJitter(double jitterFactor, int expectedMilliseconds)
    {
        var delay = ModuleRetryShieldFactory.CalculateDelay(
            retryAttempt: 2,
            baseDelay: TimeSpan.FromMilliseconds(100),
            jitterFactor);

        await Assert.That(delay).IsEqualTo(TimeSpan.FromMilliseconds(expectedMilliseconds));
    }

    [Test]
    public async Task RetryDelayCalculator_ZeroDelay_RemainsZeroAtMaximumAttempt()
    {
        var delay = ModuleRetryShieldFactory.CalculateDelay(
            retryAttempt: int.MaxValue,
            baseDelay: TimeSpan.Zero,
            jitterFactor: 0.5);

        await Assert.That(delay).IsEqualTo(TimeSpan.Zero);
    }

    [Test]
    public async Task RetryDelayCalculator_MaximumDelay_RemainsInTimeSpanRange()
    {
        var delay = ModuleRetryShieldFactory.CalculateDelay(
            retryAttempt: 2,
            baseDelay: TimeSpan.MaxValue,
            jitterFactor: 1);

        await Assert.That(delay).IsEqualTo(TimeSpan.MaxValue);
    }

    #endregion

    #region WithIgnoreFailures Tests

    [Test]
    public async Task WithIgnoreFailures_Always_SetsIgnoreFailuresCondition()
    {
        var config = ModuleConfiguration.Create()
            .WithIgnoreFailures()
            .Build();

        await Assert.That(config.IgnoreFailuresCondition).IsNotNull();

        var context = Mock.Of<IModuleContext>();
        var result = await config.IgnoreFailuresCondition!(context, new Exception("test"));

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task WithIgnoreFailuresWhen_SyncCondition_SetsIgnoreFailuresCondition()
    {
        var config = ModuleConfiguration.Create()
            .WithIgnoreFailuresWhen((ctx, ex) => ex.Message == "ignore")
            .Build();

        await Assert.That(config.IgnoreFailuresCondition).IsNotNull();

        var context = Mock.Of<IModuleContext>();

        var shouldIgnore = await config.IgnoreFailuresCondition!(context, new Exception("ignore"));
        await Assert.That(shouldIgnore).IsTrue();

        var shouldNotIgnore = await config.IgnoreFailuresCondition!(context, new Exception("fail"));
        await Assert.That(shouldNotIgnore).IsFalse();
    }

    [Test]
    public async Task WithIgnoreFailuresWhen_AsyncCondition_SetsIgnoreFailuresCondition()
    {
        var config = ModuleConfiguration.Create()
            .WithIgnoreFailuresWhen(async (ctx, ex) =>
            {
                await Task.Delay(1).ConfigureAwait(false);
                return ex.Message == "ignore";
            })
            .Build();

        var context = Mock.Of<IModuleContext>();

        var shouldIgnore = await config.IgnoreFailuresCondition!(context, new Exception("ignore"));
        await Assert.That(shouldIgnore).IsTrue();

        var shouldNotIgnore = await config.IgnoreFailuresCondition!(context, new Exception("fail"));
        await Assert.That(shouldNotIgnore).IsFalse();
    }

    #endregion

    #region WithAlwaysRun Tests

    [Test]
    public async Task WithAlwaysRun_SetsAlwaysRun()
    {
        var config = ModuleConfiguration.Create()
            .WithAlwaysRun()
            .Build();

        await Assert.That(config.AlwaysRun).IsTrue();
    }

    #endregion

    #region Fluent Chaining Tests

    [Test]
    public async Task Builder_Configures_Scheduling_Metadata_And_Dependencies()
    {
        var config = ModuleConfiguration.Create()
            .WithNotInParallel("database")
            .WithPriority(ModulePriority.Critical)
            .WithExecutionHint(ExecutionType.IoIntensive)
            .WithTags("deploy", "production")
            .WithCategory("release")
            .DependsOn<DependencyModule>()
            .Build();

        using (Assert.Multiple())
        {
            await Assert.That(config.ParallelConstraintKeys).IsEquivalentTo(new[] { "database" });
            await Assert.That(config.Priority).IsEqualTo(ModulePriority.Critical);
            await Assert.That(config.ExecutionType).IsEqualTo(ExecutionType.IoIntensive);
            await Assert.That(config.Tags).IsEquivalentTo(new[] { "deploy", "production" });
            await Assert.That(config.Category).IsEqualTo("release");
            await Assert.That(config.Dependencies.Select(dependency => dependency.ModuleType))
                .IsEquivalentTo(new[] { typeof(DependencyModule) });
        }
    }

    [Test]
    public async Task Builder_FluentChaining_AllMethodsChain()
    {
        var config = ModuleConfiguration.Create()
            .WithSkipWhen(_ => SkipDecision.DoNotSkip)
            .WithTimeout(TimeSpan.FromMinutes(1))
            .WithRetry(3)
            .WithIgnoreFailures()
            .WithAlwaysRun()
            .WithNotInParallel("shared")
            .WithPriority(ModulePriority.High)
            .WithExecutionHint(ExecutionType.CpuIntensive)
            .WithTags("build")
            .WithCategory("ci")
            .DependsOnOptional<DependencyModule>()
            .Build();

        using (Assert.Multiple())
        {
            await Assert.That((object?) config.SkipCondition).IsNotNull();
            await Assert.That(config.Timeout).IsEqualTo(TimeSpan.FromMinutes(1));
            await Assert.That(config.RetryConfiguration).IsNotNull();
            await Assert.That(config.IgnoreFailuresCondition).IsNotNull();
            await Assert.That(config.AlwaysRun).IsTrue();
            await Assert.That(config.ParallelConstraintKeys).IsEquivalentTo(new[] { "shared" });
            await Assert.That(config.Priority).IsEqualTo(ModulePriority.High);
            await Assert.That(config.ExecutionType).IsEqualTo(ExecutionType.CpuIntensive);
            await Assert.That(config.Tags).Contains("build");
            await Assert.That(config.Category).IsEqualTo("ci");
            await Assert.That(config.Dependencies).HasSingleItem();
        }
    }

    #endregion

    private static bool ContainsKevlarType(Type type) =>
        type.Namespace?.StartsWith("Kevlar", StringComparison.Ordinal) == true
        || type.GetGenericArguments().Any(ContainsKevlarType);
}

File: test/ModularPipelines.UnitTests.FSharp/Configuration/ModuleConfigurationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Configuration

open System
open System.Threading.Tasks
open ModularPipelines.Configuration
open ModularPipelines.Context
open ModularPipelines.Models
open Moq
open Polly
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type ModuleConfigurationTests() =
    [<Test>]
    member _.Default_SkipCondition_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.SkipCondition).IsNull())
    }

    [<Test>]
    member _.Default_Timeout_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.Timeout.HasValue).IsFalse())
    }

    [<Test>]
    member _.Default_RetryPolicyFactory_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.RetryPolicyFactory).IsNull())
    }

    [<Test>]
    member _.Default_IgnoreFailuresCondition_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.IgnoreFailuresCondition).IsNull())
    }

    [<Test>]
    member _.Default_AlwaysRun_IsFalse() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.AlwaysRun).IsFalse())
    }

    [<Test>]
    member _.Default_OnBeforeExecute_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.OnBeforeExecute).IsNull())
    }

    [<Test>]
    member _.Default_OnAfterExecute_IsNull() = async {
        let config = ModuleConfiguration.Default
        do! check(Assert.That(config.OnAfterExecute).IsNull())
    }

    [<Test>]
    member _.Create_ReturnsBuilder() = async {
        let builder = ModuleConfiguration.Create()
        do! check(Assert.That(builder).IsNotNull())
        do! check(Assert.That(builder).IsTypeOf<ModuleConfigurationBuilder>())
    }

    [<Test>]
    member _.WithSkipWhen_SyncBool_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(Func<bool>(fun () -> true))
                .Build()

        do! check(Assert.That(config.SkipCondition).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
    }

    [<Test>]
    member _.WithSkipWhen_SyncBoolFalse_ReturnsDoNotSkip() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(Func<bool>(fun () -> false))
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsFalse())
    }

    [<Test>]
    member _.WithSkipWhen_AsyncBool_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(
                    Func<Task<bool>>(fun () ->
                        task {
                            do! Task.Delay(1)
                            return true
                        })
                )
                .Build()

        do! check(Assert.That(config.SkipCondition).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
    }

    [<Test>]
    member _.WithSkipWhen_SyncSkipDecision_SetsSkipCondition() = async {
        let expectedDecision = SkipDecision.Skip("Test reason")

        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(Func<SkipDecision>(fun () -> expectedDecision))
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(decision.Reason), "Test reason"))
    }

    [<Test>]
    member _.WithSkipWhen_AsyncSkipDecision_SetsSkipCondition() = async {
        let expectedDecision = SkipDecision.Skip("Async reason")

        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(
                    Func<Task<SkipDecision>>(fun () ->
                        task {
                            do! Task.Delay(1)
                            return expectedDecision
                        })
                )
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(decision.Reason), "Async reason"))
    }

    [<Test>]
    member _.WithSkipWhen_WithContext_SyncBool_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(Func<IModuleContext, bool>(fun ctx -> not (obj.ReferenceEquals(ctx, null))))
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
    }

    [<Test>]
    member _.WithSkipWhen_WithContext_AsyncBool_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(
                    Func<IModuleContext, Task<bool>>(fun ctx ->
                        task {
                            do! Task.Delay(1)
                            return not (obj.ReferenceEquals(ctx, null))
                        })
                )
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
    }

    [<Test>]
    member _.WithSkipWhen_WithContext_SyncSkipDecision_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(
                    Func<IModuleContext, SkipDecision>(fun ctx ->
                        if not (obj.ReferenceEquals(ctx, null)) then
                            SkipDecision.Skip("Has context")
                        else
                            SkipDecision.DoNotSkip)
                )
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(decision.Reason), "Has context"))
    }

    [<Test>]
    member _.WithSkipWhen_WithContext_AsyncSkipDecision_SetsSkipCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(
                    Func<IModuleContext, Task<SkipDecision>>(fun ctx ->
                        task {
                            do! Task.Delay(1)

                            return
                                if not (obj.ReferenceEquals(ctx, null)) then
                                    SkipDecision.Skip("Async context")
                                else
                                    SkipDecision.DoNotSkip
                        })
                )
                .Build()

        let context = Mock.Of<IModuleContext>()
        let! decision = config.SkipCondition.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(decision.ShouldSkip).IsTrue())
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(decision.Reason), "Async context"))
    }

    [<Test>]
    member _.WithTimeout_SetsTimeout() = async {
        let timeout = TimeSpan.FromMinutes(5.0)

        let config =
            ModuleConfiguration.Create()
                .WithTimeout(timeout)
                .Build()

        do! check(Assert.That(config.Timeout.HasValue).IsTrue())
        do! check(Assert.That(config.Timeout.Value = timeout).IsTrue())
    }

    [<Test>]
    member _.WithRetryPolicy_Direct_SetsRetryPolicyFactory() = async {
        let policy = Policy.NoOpAsync()

        let config =
            ModuleConfiguration.Create()
                .WithRetryPolicy(policy)
                .Build()

        do! check(Assert.That(config.RetryPolicyFactory).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        let result = config.RetryPolicyFactory.Invoke(context)

        do! check(Assert.That(obj.ReferenceEquals(result, policy)).IsTrue())
    }

    [<Test>]
    member _.WithRetryPolicy_Factory_SetsRetryPolicyFactory() = async {
        let policy = Policy.NoOpAsync()

        let config =
            ModuleConfiguration.Create()
                .WithRetryPolicy(Func<IModuleContext, IAsyncPolicy>(fun _ -> policy))
                .Build()

        do! check(Assert.That(config.RetryPolicyFactory).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        let result = config.RetryPolicyFactory.Invoke(context)

        do! check(Assert.That(obj.ReferenceEquals(result, policy)).IsTrue())
    }

    [<Test>]
    member _.WithRetryCount_SetsRetryPolicyFactory() = async {
        let config =
            ModuleConfiguration.Create()
                .WithRetryCount(3)
                .Build()

        do! check(Assert.That(config.RetryPolicyFactory).IsNotNull())
    }

    [<Test>]
    member _.WithIgnoreFailures_Always_SetsIgnoreFailuresCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithIgnoreFailures()
                .Build()

        do! check(Assert.That(config.IgnoreFailuresCondition).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        let! result = config.IgnoreFailuresCondition.Invoke(context, Exception("test")) |> Async.AwaitTask

        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.WithIgnoreFailuresWhen_SyncCondition_SetsIgnoreFailuresCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithIgnoreFailuresWhen(Func<IModuleContext, Exception, bool>(fun _ ex -> ex.Message = "ignore"))
                .Build()

        do! check(Assert.That(config.IgnoreFailuresCondition).IsNotNull())

        let context = Mock.Of<IModuleContext>()

        let! shouldIgnore = config.IgnoreFailuresCondition.Invoke(context, Exception("ignore")) |> Async.AwaitTask
        do! check(Assert.That(shouldIgnore).IsTrue())

        let! shouldNotIgnore = config.IgnoreFailuresCondition.Invoke(context, Exception("fail")) |> Async.AwaitTask
        do! check(Assert.That(shouldNotIgnore).IsFalse())
    }

    [<Test>]
    member _.WithIgnoreFailuresWhen_AsyncCondition_SetsIgnoreFailuresCondition() = async {
        let config =
            ModuleConfiguration.Create()
                .WithIgnoreFailuresWhen(
                    Func<IModuleContext, Exception, Task<bool>>(fun _ ex ->
                        task {
                            do! Task.Delay(1)
                            return ex.Message = "ignore"
                        })
                )
                .Build()

        let context = Mock.Of<IModuleContext>()

        let! shouldIgnore = config.IgnoreFailuresCondition.Invoke(context, Exception("ignore")) |> Async.AwaitTask
        do! check(Assert.That(shouldIgnore).IsTrue())

        let! shouldNotIgnore = config.IgnoreFailuresCondition.Invoke(context, Exception("fail")) |> Async.AwaitTask
        do! check(Assert.That(shouldNotIgnore).IsFalse())
    }

    [<Test>]
    member _.WithAlwaysRun_SetsAlwaysRun() = async {
        let config =
            ModuleConfiguration.Create()
                .WithAlwaysRun()
                .Build()

        do! check(Assert.That(config.AlwaysRun).IsTrue())
    }

    [<Test>]
    member _.WithBeforeExecute_SetsOnBeforeExecute() = async {
        let mutable executed = false

        let config =
            ModuleConfiguration.Create()
                .WithBeforeExecute(
                    Func<IModuleContext, Task>(fun _ ->
                        task {
                            do! Task.Delay(1)
                            executed <- true
                        })
                )
                .Build()

        do! check(Assert.That(config.OnBeforeExecute).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        do! config.OnBeforeExecute.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(executed).IsTrue())
    }

    [<Test>]
    member _.WithAfterExecute_SetsOnAfterExecute() = async {
        let mutable executed = false

        let config =
            ModuleConfiguration.Create()
                .WithAfterExecute(
                    Func<IModuleContext, Task>(fun _ ->
                        task {
                            do! Task.Delay(1)
                            executed <- true
                        })
                )
                .Build()

        do! check(Assert.That(config.OnAfterExecute).IsNotNull())

        let context = Mock.Of<IModuleContext>()
        do! config.OnAfterExecute.Invoke(context) |> Async.AwaitTask

        do! check(Assert.That(executed).IsTrue())
    }

    [<Test>]
    member _.Builder_FluentChaining_AllMethodsChain() = async {
        let policy = Policy.NoOpAsync()
        let timeout = TimeSpan.FromMinutes(1.0)

        let config =
            ModuleConfiguration.Create()
                .WithSkipWhen(Func<bool>(fun () -> false))
                .WithTimeout(timeout)
                .WithRetryPolicy(policy)
                .WithIgnoreFailures()
                .WithAlwaysRun()
                .WithBeforeExecute(Func<IModuleContext, Task>(fun _ -> Task.CompletedTask))
                .WithAfterExecute(Func<IModuleContext, Task>(fun _ -> Task.CompletedTask))
                .Build()

        do! check(Assert.That(config.SkipCondition).IsNotNull())
        do! check(Assert.That(config.Timeout.HasValue).IsTrue())
        do! check(Assert.That(config.Timeout.Value = timeout).IsTrue())
        do! check(Assert.That(config.RetryPolicyFactory).IsNotNull())
        do! check(Assert.That(config.IgnoreFailuresCondition).IsNotNull())
        do! check(Assert.That(config.AlwaysRun).IsTrue())
        do! check(Assert.That(config.OnBeforeExecute).IsNotNull())
        do! check(Assert.That(config.OnAfterExecute).IsNotNull())
    }
